products/data-schema-azuresignalrservice.md

```diff
@@ -0,0 +1,750 @@
+- **Kusto Cluster**:
+signalrinsights.eastus2
+- **Kusto Database**:
+SignalRBI
+- **Access Control**:
+**[TODO] Data Engineer**: If this product’s data has high confidentiality concerns, please specify the allowed **groups/users** here. If left blank, general users will be permitted to run analyses on this product, including cross-product scenarios.  
+
+-----
+
+# Azure SignalR Service + Web PubSub Kusto Query Playbook
+
+## Overview
+- Product: AzureSignalRService (covers both Azure SignalR Service and Azure Web PubSub)
+- Summary: Power Query M blocks defining Azure Data Explorer (KQL) queries for Azure SignalR Service and Azure Web PubSub, covering QoS (operation/connectivity), service usage, billing/revenue, subscriptions/customers retention and category flows, features usage (including SocketIO, AutoScale, Aspire), runtime/ARM/VS telemetry, and error diagnostics. Main execution target is the signalrinsights.eastus2 cluster and SignalRBI database.
+- Primary cluster/database:
+  - Cluster: signalrinsights.eastus2
+  - Database: SignalRBI
+- Alternates observed:
+  - ddtelinsights (for WebTools_Raw developer telemetry)
+  - armprodgbl.eastus (ARM Requests/HttpIncomingRequests; typically accessed via a union helper)
+
+## Conventions & Assumptions
+
+### Time semantics
+- Observed timestamp columns by source:
+  - Operation/Connectivity QoS: env_time (datetime); often binned to day/week/month
+  - Daily fact tables (usage/billing/feature/runtime): Date (datetime)
+  - Hourly errors/exceptions: DateHour (string/datetime-like), Timestamp (datetime)
+  - Developer telemetry (WebTools_Raw): AdvancedServerTimestampUtc (datetime)
+- Canonical Time Column: Date
+  - Align non-Date tables to Date with explicit aliasing:
+    - QoS tables: project-rename Date = env_time
+    - Hourly sources: project Date = startofday(Timestamp) or Date = startofday(todatetime(DateHour))
+    - Developer telemetry: project Date = startofday(AdvancedServerTimestampUtc)
+- Typical windows and binning
+  - ago windows: 10d, 30d, 43d, 60d, 61d
+  - Week bins: startofweek(Date)
+  - Month bins: startofmonth(Date)
+  - Day bins: startofday(...)
+- Timezone: Not specified; assume UTC.
+
+### Freshness & Completeness Gates
+- Daily ingestion can lag; current day may be incomplete. Queries commonly gate to a last “complete” day using sentinel slices (regions).
+- Multi-slice completeness gate pattern (observed):
+  - Require ≥N regions to report for the day, then pick the latest such day.
+  - Example patterns:
+    - Using two regions:
+      kusto
+      let EndDate = toscalar(
+        BillingUsage_Daily
+        | where Date > ago(61d) and Region in ('australiaeast', 'westeurope')
+        | summarize counter = dcount(Region) by Date
+        | where counter >= 2
+        | top 1 by Date desc
+        | project Date
+      );
+    - Using broad estate (e.g., ≥38 regions):
```

Comment:
not valid now, better to have way to ignore some content with similar function but low frequent use, which seems to be a signal that is out of date.